### PR TITLE
[READY] Add RefactorRename support to typescript

### PR DESCRIPTION
**Pull request by puremourning**
Fixes https://github.com/Valloric/YouCompleteMe/issues/2018

This also contains a pretty important fix to the javascript tests where the tests weren't actually executing a matcher on `RefactorRename` tests.

Currently RFC because this was done pretty quickly and tested fairly casually. I'm pretty confident it works because it is almost identical to the javascript implementation (In the sense that the hard work is still done by the client, which is unchanged), but still worth some additional testing.

[Review on Reviewable](https://reviewable.io/reviews/valloric/ycmd/417)

### CODE REVIEW

**Comment by Valloric** [comment]
Thanks!

:lgtm_strong: with minor comment.

---

Review status: 0 of 8 files reviewed at latest revision, 1 unresolved discussion, some commit checks failed.

---

*[ycmd/completers/typescript/typescript_completer.py, line 445 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCEQIV6HzKC3kIqv_JP:-KCEQIV6HzKC3kIqv_JQ:1501001436) ([raw file](https://github.com/valloric/ycmd/blob/40d2ebb6d1e4d98fa1867d1c003b25ede1c8b1e1/ycmd/completers/typescript/typescript_completer.py#L445)):*
This list comprehension is a bit too clever; unwrapping it into a for loop would improve readability.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/417#-:-KCEQYg-QGtwhcIsReF8:-2001772269)*

File: ycmd/completers/typescript/typescript_completer.py [hunk 1]
```diff
@@ -377,6 +380,76 @@ def _GetDoc( self, request_data ):
...
+                                   request_data['filepath'] )
+
+    chunks = []
+    for file_replacement in response[ 'locs' ]:
+      chunks.extend( [
+        responses.FixItChunk(
+          new_name,
+          responses.Range(
+            responses.Location( l[ 'start'][ 'line' ],
+                                l[ 'start'][ 'offset' ],
```
Comment (icholy) [inline ycmd/completers/typescript/typescript_completer.py]: Missing space after `'start'`

Comment (puremourning) [inline ycmd/completers/typescript/typescript_completer.py]: Nice catch, thanks.

**Comment by icholy** [comment]
Review status: 0 of 8 files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[ycmd/completers/typescript/typescript_completer.py, line 430 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCHhgc_KEGgducFVt2E:-KCHhgcaYikUx91Ahstu:-264081020) ([raw file](https://github.com/valloric/ycmd/blob/40d2ebb6d1e4d98fa1867d1c003b25ede1c8b1e1/ycmd/completers/typescript/typescript_completer.py#L430)):*
Spaces missing around `'line_num'`, `'column_num'` & `'filepath'`

---

*[ycmd/completers/typescript/typescript_completer.py, line 445 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCEQIV6HzKC3kIqv_JP:-KCHi0vOi7mkHS81kA5D:1065261914) ([raw file](https://github.com/valloric/ycmd/blob/40d2ebb6d1e4d98fa1867d1c003b25ede1c8b1e1/ycmd/completers/typescript/typescript_completer.py#L445)):*
Might be useful to pull it out into a function similar to `_ConvertCompletionData` and `_ConvertDetailsCompletionData`

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/417)*

**Comment by vheon** [comment]
Reviewed 8 of 8 files at r1.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks failed.

---

*[ycmd/tests/javascript/subcommands_test.py, line 386 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCHtVtlR4A_yFrgyFFX:-KCHtVtlR4A_yFrgyFFY:1511200219) ([raw file](https://github.com/valloric/ycmd/blob/40d2ebb6d1e4d98fa1867d1c003b25ede1c8b1e1/ycmd/tests/javascript/subcommands_test.py#L386)):*
Why did we change this subcommand name? I kinda liked the `FixIt` prefix :smirk: 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/417)*

**Comment by puremourning** [comment]
Grumble.

It seems that the paths returned by `tsserver` on windows are of the form `C:/dir/dir/dir/file` whereas all other paths are of the form `C:\\dir\\dir\\dir\\file` so all of the tests are failing.

`os.path.normpath` looks to be the solution (its docs says that it does this), so I'm punting on that, rather than booting my windows VM.

---

Review status: 7 of 8 files reviewed at latest revision, 4 unresolved discussions.

---

*[ycmd/completers/typescript/typescript_completer.py, line 430 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCHhgc_KEGgducFVt2E:-KCHuSEz4txhFrDTl3Df:-499146009) ([raw file](https://github.com/valloric/ycmd/blob/40d2ebb6d1e4d98fa1867d1c003b25ede1c8b1e1/ycmd/completers/typescript/typescript_completer.py#L430)):*
Done.

---

*[ycmd/completers/typescript/typescript_completer.py, line 445 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCEQIV6HzKC3kIqv_JP:-KCHuPhjJDvV8P6Rmyl2:2102527246) ([raw file](https://github.com/valloric/ycmd/blob/40d2ebb6d1e4d98fa1867d1c003b25ede1c8b1e1/ycmd/completers/typescript/typescript_completer.py#L445)):*
I refactored it

---

*[ycmd/tests/javascript/subcommands_test.py, line 386 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCHtVtlR4A_yFrgyFFX:-KCHuGna5GkS7qr4AaZ9:-612025039) ([raw file](https://github.com/valloric/ycmd/blob/40d2ebb6d1e4d98fa1867d1c003b25ede1c8b1e1/ycmd/tests/javascript/subcommands_test.py#L386)):*
I know, but the consensus was to use Refactor. This PR doesn't change the name, it actually just fixes the (previously outrageously broken) tests.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/417#-:-KCHwH-INoacSZDuJbsz:-1850111349)*

**Comment by vheon** [comment]
:lgtm: assuming the tests passing :smiley: 

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[ycmd/tests/javascript/subcommands_test.py, line 386 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCHtVtlR4A_yFrgyFFX:-KCI19EXqITMwNlbWTZB:1095018191) ([raw file](https://github.com/valloric/ycmd/blob/40d2ebb6d1e4d98fa1867d1c003b25ede1c8b1e1/ycmd/tests/javascript/subcommands_test.py#L386)):*
I see... well I'll live :unamused: 

Why the tests weren't failing though? I mean we were sending a non existing command and our assertions didn't notice that? :confused: 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/417#-:-KCI1eclFzRXcGXaprC-:-2081290708)*

**Comment by puremourning** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[ycmd/tests/javascript/subcommands_test.py, line 386 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCHtVtlR4A_yFrgyFFX:-KCI21D8O7OFzX2rDkdj:1153132950) ([raw file](https://github.com/valloric/ycmd/blob/40d2ebb6d1e4d98fa1867d1c003b25ede1c8b1e1/ycmd/tests/javascript/subcommands_test.py#L386)):*
OK fine, you force me to admit my stupidity and don my sackcloth and ashes.

Basically all the tests were doing

```python
  'data': {
    'stuff: matcher()
  }
```

But the check was doing:

```python
assert_that( actual_data, thing[ 'data' ] )
```

So it was actually _checking_ **nothing**. Oh the shame. By changing to:

```python
  'data': has_items( {
    'stuff': matcher()
  } )
```

we now, *actually* check the responses.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/417)*

**Comment by vheon** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[ycmd/tests/javascript/subcommands_test.py, line 386 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCHtVtlR4A_yFrgyFFX:-KCI4-TZiOjLEcrV4dQt:1347913041) ([raw file](https://github.com/valloric/ycmd/blob/40d2ebb6d1e4d98fa1867d1c003b25ede1c8b1e1/ycmd/tests/javascript/subcommands_test.py#L386)):*
Sorry :stuck_out_tongue: but thanks for the explanation :+1: 

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/417)*

**Comment by puremourning** [comment]
Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---

*[ycmd/tests/javascript/subcommands_test.py, line 386 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCHtVtlR4A_yFrgyFFX:-KCI4BDEIK5TDauqfvuM:-1327967420) ([raw file](https://github.com/valloric/ycmd/blob/40d2ebb6d1e4d98fa1867d1c003b25ede1c8b1e1/ycmd/tests/javascript/subcommands_test.py#L386)):*
'sok. I already came clean on gitter. It was only a matter of time before my deception of competence became public knowledge :)

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/417)*

**Comment by puremourning** [comment]
OK

@puremourning boots his boot camp 

---

Review status: all files reviewed at latest revision, 2 unresolved discussions, some commit checks pending.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/417#-:-KCI6-ZITpRURE38NAJ8:744677844)*

**Comment by micbou** [comment]
Reviewed 7 of 8 files at r1, 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 3 unresolved discussions, some commit checks pending.

---

*[ycmd/tests/test_utils.py, line 178 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCI4ieqZgruE55eiIye:-KCI4ieqZgruE55eiIyf:-2023835775) ([raw file](https://github.com/valloric/ycmd/blob/8c395cb0f559bed38a9ee0e674316ba5911f7137/ycmd/tests/test_utils.py#L178)):*
Could we move them with the other matchers of this file?

---

*[ycmd/tests/javascript/subcommands_test.py, line 386 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCHtVtlR4A_yFrgyFFX:-KCI58zDZL-kqpZtRdWj:1936423648) ([raw file](https://github.com/valloric/ycmd/blob/40d2ebb6d1e4d98fa1867d1c003b25ede1c8b1e1/ycmd/tests/javascript/subcommands_test.py#L386)):*
That's funny. I was working on adding this subcommand to the Python completer (it would have be awesome for refactoring the tests) and since I am lazy, I copied one of the test from Javascript and didn't understand why it was passing without any change. I thought it was a bug introduced by my tests refactoring. Glad it is not the case.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/417)*

**Comment by puremourning** [comment]
OK looks like the tests are all passing now, and I've done a load more testing. So I'm happy with this. Changed to [READY]

**Comment by Valloric** [comment]
Reviewed 5 of 8 files at r1, 3 of 3 files at r4.
Review status: all files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/completers/typescript/typescript_completer.py, line 434 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCJ07DJhinW8F7ZspIS:-KCJ07DJhinW8F7ZspIT:-1826066479) ([raw file](https://github.com/valloric/ycmd/blob/c36cb51b4ab4656d1364e52f822d443bf319edfe/ycmd/completers/typescript/typescript_completer.py#L434)):*
I'm not actually a big fan of nested functions since they're worse encapsulation-wise than free functions. The reason they're worse is because they have access to all the vars of the enclosing function, and the only way to verify they don't do funky  mutaty stuff with that data is to go over them line by line. 

Nested functions can be fine when they're very small and when they + their outer function taken together are fairly small too. This doesn't pass that test for me.

So I'd rather see these extracted as free functions, since IMO that improves readability.

---

*[ycmd/tests/javascript/subcommands_test.py, line 386 \[r1\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCHtVtlR4A_yFrgyFFX:-KCJ0odpZN3Sja1SYBf5:1533248332) ([raw file](https://github.com/valloric/ycmd/blob/40d2ebb6d1e4d98fa1867d1c003b25ede1c8b1e1/ycmd/tests/javascript/subcommands_test.py#L386)):*
@puremourning psh, I do things that are at least 2x dumber several times a week. This mistake is pocket change. :)

Last year at Google I spent 3 weeks of my time and occupied 10k machines for the duration researching a ranking function only to discover that all the fancy signals I was throwing at it weren't doing shit because the last stage of the MapReduce pipeline was sorting on an uninitialized integer. You know, the stage that's "so trivial it doesn't need any tests."

And that's nowhere _near_ the dumbest thing I did in the four years I was there.

"Only people who do no work make no mistakes."

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/417)*

**Comment by puremourning** [comment]
Review status: 6 of 8 files reviewed at latest revision, 3 unresolved discussions.

---

*[ycmd/completers/typescript/typescript_completer.py, line 434 \[r4\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCJ07DJhinW8F7ZspIS:-KCXKTJ8RtiqqIg-9YPg:557666262) ([raw file](https://github.com/valloric/ycmd/blob/c36cb51b4ab4656d1364e52f822d443bf319edfe/ycmd/completers/typescript/typescript_completer.py#L434)):*
Fair point. I often like the whole of a method's logic to be visible in one place rather than having to jump around in the editor, but I agree that this is better.

Done.

---

*[ycmd/tests/test_utils.py, line 178 \[r3\]](https://reviewable.io:443/reviews/valloric/ycmd/417#-KCI4ieqZgruE55eiIye:-KCI8XBGnK0hwuOffl4J:22359712) ([raw file](https://github.com/valloric/ycmd/blob/8c395cb0f559bed38a9ee0e674316ba5911f7137/ycmd/tests/test_utils.py#L178)):*
sure. done.

---


*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/417)*

**Comment by Valloric** [comment]
Thanks!

@homu r+

---

Review status: 6 of 8 files reviewed at latest revision, 2 unresolved discussions.

---



*Comments from the [review on Reviewable.io](https://reviewable.io:443/reviews/valloric/ycmd/417#-:-KCXLooK8MR2el5KWtQr:-388066883)*
